### PR TITLE
Feat: Funding Pot V1 - Creating and Editing a Round

### PR DESCRIPTION
**Pull request by zzzuhaibmohd**
### Overview
This pull request is for the review of the `Funding_Pot_v1` contracts after adding implementation of `createRound()`, `editRound()` and `setAccessCriteriaForRound()`

### Key Changes
- **Funding Pot Admin:** Added logic to allow only the `FUNDING_POT_ADMIN_ROLE` to call `createRound()`, `editRound()` and `setAccessCriteriaForRound()`
- **Input Validation:** Added various custom errors to handle incorrect input validation and revert in case invalid input is passed
- **Unit Tests:**
  - Added unit and fuzz tests for `createRound()`, `editRound()` and `setAccessCriteriaForRound()` following the gherkin style method
- **Code Cleanup:** Removed unnecessary state variables and outdated templates to streamline the codebase.

### Notable Commits
- **Mar 17, 2025:** Create and Edit Round Implementation.
- **Mar 18, 2025:** Add tests related to create/edit round functionality.
- **Mar 24, 2025:** Implement the access control criteria 
- **Mar 25, 2025:** Add tests related to access control criteria.

### Focus Areas
- Provide feedback on implementation, unit tests and identifying any blind spots or areas where we can improve.
- If missing to handle any edge case or introduced a bug in our current implementation!

### CODE REVIEW

File: src/modules/logicModule/interfaces/ILM_PC_FundingPot_v1.sol [hunk 1]
```diff
@@ -6,15 +6,235 @@ import {IERC20PaymentClientBase_v2} from
     "@lm/interfaces/IERC20PaymentClientBase_v2.sol";
 
 interface ILM_PC_FundingPot_v1 is IERC20PaymentClientBase_v2 {
-// -------------------------------------------------------------------------
-// Events
+    //--------------------------------------------------------------------------
```
Comment (fabianschu) [inline src/modules/logicModule/interfaces/ILM_PC_FundingPot_v1.sol]:
```suggestion
    // -------------------------------------------------------------------------
```

File: src/modules/logicModule/interfaces/ILM_PC_FundingPot_v1.sol [hunk 2]
```diff
@@ -6,15 +6,235 @@ import {IERC20PaymentClientBase_v2} from
...
+        address[] allowedAddresses; // Explicit allowlist
+    }
 
-// -------------------------------------------------------------------------
-// Public - Mutating
+    // -------------------------------------------------------------------------
+    // Enums
+
+    /// @notice Enum used to identify the type of access criteria.
+    enum AccessCriteriaId {
```
Comment (fabianschu) [inline src/modules/logicModule/interfaces/ILM_PC_FundingPot_v1.sol]: Can we rename this to `AccessCriteriaType` cause the duplicate use of `id` in the ` struct Round` as well as in the `struct AccessCriteria` can be confusing?

File: src/modules/logicModule/interfaces/ILM_PC_FundingPot_v1.sol [hunk 3]
```diff
@@ -6,15 +6,235 @@ import {IERC20PaymentClientBase_v2} from
...
+        OPEN, // 0
```
Comment (fabianschu) [inline src/modules/logicModule/interfaces/ILM_PC_FundingPot_v1.sol]: I feel a bit uncomfortable with the default value of the enum being `OPEN`. If the `AccessCriteria` is not set for an `id` in the `Round`, then I think a check for that `AccessCriteriaType` would make it look like as if the (nonexistent) AccessCriteria is set to OPEN. 

I'm sure, additional checks can/will prevent any undesired consequences, but putting OPEN in a different slot in the enum would play it safer, or putting `UNSET` explicitly in there at slot zero. 

Maybe I'm thinking too far ahead and this isn't really a problem, so please feel free to ignore, if you disagree :)

File: src/modules/logicModule/interfaces/ILM_PC_FundingPot_v1.sol [hunk 4]
```diff
@@ -6,15 +6,235 @@ import {IERC20PaymentClientBase_v2} from
...
+    // Events
+
+    /// @notice Emitted when a new round is created.
+    /// @dev    This event signals the creation of a new round with specific parameters.
+    /// @param  roundId_ The unique identifier for the round.
+    /// @param  roundStart_ The timestamp when the round starts.
+    /// @param  roundEnd_ The timestamp when the round ends.
+    /// @param  roundCap_ The maximum allocation or cap for the round.
+    /// @param  hookContract_ The address of an optional hook contract for custom logic.
+    /// @param  closureMechanism_ A boolean indicating whether a specific closure mechanism is enabled.
```
Comment (fabianschu) [inline src/modules/logicModule/interfaces/ILM_PC_FundingPot_v1.sol]: I think the naming of this param could be more specific, like `autoClosure_` or similar.

File: src/modules/logicModule/interfaces/ILM_PC_FundingPot_v1.sol [hunk 5]
```diff
@@ -6,15 +6,235 @@ import {IERC20PaymentClientBase_v2} from
...
+        bool globalAccumulativeCaps_
+    );
+
+    /// @notice Emitted when access criteria is set for a round.
+    /// @param  roundId_ The unique identifier of the round.
+    /// @param  accessId_ The identifier of the access criteria.
+    /// @param  accessCriteria_ The access criteria.
+    event AccessCriteriaSet(
+        uint64 indexed roundId_, uint8 accessId_, AccessCriteria accessCriteria_
+    );
```
Comment (fabianschu) [inline src/modules/logicModule/interfaces/ILM_PC_FundingPot_v1.sol:108]: For the Round, there are two separate events for creation and editing. Can we do the same for `AccessCriteria`?

File: src/modules/logicModule/interfaces/ILM_PC_FundingPot_v1.sol [hunk 6]
```diff
@@ -6,15 +6,235 @@ import {IERC20PaymentClientBase_v2} from
...
+    error Module__LM_PC_FundingPot__HookFunctionRequiredWithContract();
+
+    /// @notice Thrown when a hook contract is specified without a hook function.
+    error Module__LM_PC_FundingPot__HookFunctionRequiredWithHookContract();
+
+    /// @notice Thrown when a hook function is specified without a hook contract.
+    error Module__LM_PC_FundingPot__HookContractRequiredWithHookFunction();
+
+    /// @notice Round does not exist
+    error Module__LM_PC_FundingPot__RoundNotCreated();
```
Comment (fabianschu) [inline src/modules/logicModule/interfaces/ILM_PC_FundingPot_v1.sol]: Is this a duplicate to Module__LM_PC_FundingPot__RoundDoesNotExist ?

File: src/modules/logicModule/interfaces/ILM_PC_FundingPot_v1.sol [hunk 7]
```diff
@@ -6,15 +6,235 @@ import {IERC20PaymentClientBase_v2} from
...
+        bool globalAccumulativeCaps;
+        mapping(uint64 id => AccessCriteria) accessCriterias;
+    }
 
-// -------------------------------------------------------------------------
-// Public - Getters
+    /// @notice Struct used to store information about a funding round's access criteria.
+    /// @param  nftContract Address of the NFT contract.
+    /// @param  merkleRoot Merkle root for the access criteria.
+    /// @param  allowedAddresses Mapping of addresses to their access status.
```
Comment (fabianschu) [inline src/modules/logicModule/interfaces/ILM_PC_FundingPot_v1.sol:36]: Right now this is an array. I kind of preferred the struct because it shifts the gas costs to the admin and not to the user and globally requires less gas. Wdyt?

File: src/modules/logicModule/interfaces/ILM_PC_FundingPot_v1.sol [hunk 8]
```diff
@@ -6,15 +6,235 @@ import {IERC20PaymentClientBase_v2} from
...
+    error Module__LM_PC_FundingPot__RoundEndMustBeAfterStart();
+
+    /// @notice Round has already started and cannot be modified
+    error Module__LM_PC_FundingPot__RoundAlreadyStarted();
+
+    /// @notice Hook function is required when a hook contract is provided
+    error Module__LM_PC_FundingPot__HookFunctionRequiredWithContract();
+
+    /// @notice Thrown when a hook contract is specified without a hook function.
+    error Module__LM_PC_FundingPot__HookFunctionRequiredWithHookContract();
```
Comment (fabianschu) [inline src/modules/logicModule/interfaces/ILM_PC_FundingPot_v1.sol:140]: This is a duplicate.

File: src/modules/logicModule/interfaces/ILM_PC_FundingPot_v1.sol [hunk 9]
```diff
@@ -6,15 +6,235 @@ import {IERC20PaymentClientBase_v2} from
     "@lm/interfaces/IERC20PaymentClientBase_v2.sol";
 
 interface ILM_PC_FundingPot_v1 is IERC20PaymentClientBase_v2 {
```
Comment (fabianschu) [inline src/modules/logicModule/interfaces/ILM_PC_FundingPot_v1.sol:8]: Each NatSpec comment should end with a `.` :) 

Just adding this here once, but it is relevant in many places.

File: src/modules/logicModule/interfaces/ILM_PC_FundingPot_v1.sol [hunk 10]
```diff
@@ -6,15 +6,235 @@ import {IERC20PaymentClientBase_v2} from
...
+            uint roundCap_,
+            address hookContract_,
+            bytes memory hookFunction_,
+            bool closureMechanism_,
+            bool globalAccumulativeCaps_
+        );
+
+    /// @notice Retrieves the access criteria for a specific funding round.
+    /// @param  roundId_ The unique identifier of the round to retrieve.
+    /// @param  id_ The identifier of the access criteria to retrieve.
```
Comment (fabianschu) [inline src/modules/logicModule/interfaces/ILM_PC_FundingPot_v1.sol]: For consistency, let's call this `accessCriteriaId_` and as commented elsewhere what is now `accessCriteriaId` to `accessCriteriaType`?

File: src/modules/logicModule/interfaces/ILM_PC_FundingPot_v1.sol [hunk 11]
```diff
@@ -6,15 +6,235 @@ import {IERC20PaymentClientBase_v2} from
...
+    /// @param  globalAccumulativeCaps_ A boolean indicating whether global accumulative caps are enforced.
+    event RoundEdited(
+        uint indexed roundId_,
+        uint roundStart_,
+        uint roundEnd_,
+        uint roundCap_,
+        address hookContract_,
+        bool closureMechanism_,
+        bool globalAccumulativeCaps_
+    );
```
Comment (fabianschu) [inline src/modules/logicModule/interfaces/ILM_PC_FundingPot_v1.sol]: Should we add the hookFunction here?

File: src/modules/logicModule/LM_PC_FundingPot_v1.sol [hunk 12]
```diff
@@ -97,9 +125,223 @@ contract LM_PC_FundingPot_v1 is
...
+                accessCriteria.allowedAddresses
+            );
+        } else {
+            return (
+                false,
+                accessCriteria.nftContract,
+                accessCriteria.merkleRoot,
+                accessCriteria.allowedAddresses
+            );
+        }
```
Comment (fabianschu) [inline src/modules/logicModule/LM_PC_FundingPot_v1.sol]: Maybe simplify this?

```
isOpen = accessCriteria.accessCriteriaId == AccessCriteriaId.OPEN;
return (
    isOpen,
    accessCriteria.nftContract,
    accessCriteria.merkleRoot,
    accessCriteria.allowedAddresses
);
```

File: src/modules/logicModule/interfaces/ILM_PC_FundingPot_v1.sol [hunk 13]
```diff
@@ -6,15 +6,235 @@ import {IERC20PaymentClientBase_v2} from
...
+        address hookContract_,
+        bytes memory hookFunction_,
+        bool closureMechanism_,
+        bool globalAccumulativeCaps_
+    ) external;
+
+    /// @notice Set Access Control Check
+    /// @dev    Only callable by funding pot admin and only before the round has started
+    /// @param  roundId_ ID of the round
+    /// @param  accessId_ ID of the access criteria
```
Comment (fabianschu) [inline src/modules/logicModule/interfaces/ILM_PC_FundingPot_v1.sol]: I think this should be accessCriteriaId_

File: src/modules/logicModule/interfaces/ILM_PC_FundingPot_v1.sol (hunk 9 above)
Comment (fabianschu) [reply src/modules/logicModule/interfaces/ILM_PC_FundingPot_v1.sol:8]: Also, all return values should be named as per our coding standards (will help us to automate documentation generation). If we missed this in the template contract, please excuse that.

File: src/modules/logicModule/interfaces/ILM_PC_FundingPot_v1.sol [hunk 14]
```diff
@@ -6,15 +6,235 @@ import {IERC20PaymentClientBase_v2} from
...
+    /// @param  accessCriteria_ Access criteria to set
+    function setAccessCriteriaForRound(
```
Comment (fabianschu) [inline src/modules/logicModule/interfaces/ILM_PC_FundingPot_v1.sol:256]: I think it would be more consistent to have a function for creation and one for editing, like we have for rounds. Wdyt?

Comment (fabianschu) [reply src/modules/logicModule/interfaces/ILM_PC_FundingPot_v1.sol:256]: related to this: https://github.com/InverterNetwork/contracts/pull/732#discussion_r2017269621

Comment (fabianschu) [reply src/modules/logicModule/interfaces/ILM_PC_FundingPot_v1.sol:256]: Should we then have an incrementing counter for accessCriteriaId as well?

File: src/modules/logicModule/LM_PC_FundingPot_v1.sol [hunk 15]
```diff
@@ -97,9 +125,223 @@ contract LM_PC_FundingPot_v1 is
...
+    /// @dev    Reverts if the round parameters are invalid.
+    function _validateRoundParameters(Round storage round_) internal view {
+        // Validate round start time is in the future
+        // @note: The below condition wont allow _roundStart == block.timestamp
+        if (round_.roundStart <= block.timestamp) {
+            revert Module__LM_PC_FundingPot__RoundStartMustBeInFuture();
+        }
+
+        // Validate that either end time or cap is set
+        if (round_.roundEnd == 0 || round_.roundCap == 0) {
```
Comment (fabianschu) [inline src/modules/logicModule/LM_PC_FundingPot_v1.sol]: Should this not be an `&&`?

File: src/modules/logicModule/LM_PC_FundingPot_v1.sol [hunk 16]
```diff
@@ -97,9 +125,223 @@ contract LM_PC_FundingPot_v1 is
...
+            revert Module__LM_PC_FundingPot__RoundMustHaveEndTimeOrCap();
+        }
+
+        // If end time is set, validate it's after start time
+        if (round_.roundEnd > 0 && round_.roundEnd <= round_.roundStart) {
```
Comment (fabianschu) [inline src/modules/logicModule/LM_PC_FundingPot_v1.sol]: Should we not have `<=` rather be `<`? I don't see the case where end == start such that the round is immediately closed. Maybe I'm missing something.

File: test/modules/logicModule/LM_PC_FundingPot_v1.t.sol
Comment (fabianschu) [inline test/modules/logicModule/LM_PC_FundingPot_v1.t.sol:1]: Could you remove the role constant in the beginning of the file?

File: test/modules/logicModule/LM_PC_FundingPot_v1.t.sol [hunk 17]
```diff
@@ -95,16 +92,1004 @@ contract LM_PC_FundingPot_v1_Test is ModuleTest {
...
+    │   │       └── Then it should revert
+    ├── And hook contract is set but hook function is not set
+    │   └── When user attempts to create a round
+    │       └── Then it should revert
+    ├── And hook function is set but hook contract is not set
+    │   └── When user attempts to create a round
+    │       └── Then it should revert
+    └── Given all the valid parameters are provided
+        └── When user attempts to create a round
+            └── Then it should not be active and should return the round id
```
Comment (fabianschu) [inline test/modules/logicModule/LM_PC_FundingPot_v1.t.sol]:
```suggestion
├── Given user does not have FUNDING_POT_ADMIN_ROLE
│   └── When user attempts to create a round
│       └── Then it should revert
│
└── Given user has FUNDING_POT_ADMIN_ROLE
    ├── And round start < block.timestamp
    │   └── When user attempts to create a round
    │       └── Then it should revert
    │
    ├── And round end time == 0
    ├── And round cap == 0
    │   └── When user attempts to create a round
    │       └── Then it should revert
    │
    ├── And round end time is set
    ├── And round end != 0
    ├── And round end < round start
    │   └── When user attempts to create a round
    │       └── Then it should revert
    │
    ├── And hook contract is set but hook function is not set
    │   └── When user attempts to create a round
    │       └── Then it should revert
    │
    ├── And hook function is set but hook contract is not set
    │   └── When user attempts to create a round
    │       └── Then it should revert
    │
    └── And all the valid parameters are provided
        └── When user attempts to create a round
            └── Then it should not be active and should return the round id
```

File: test/modules/logicModule/LM_PC_FundingPot_v1.t.sol [hunk 18]
```diff
@@ -95,16 +92,1004 @@ contract LM_PC_FundingPot_v1_Test is ModuleTest {
...
+        assertEq(nftContract, accessCriteria.nftContract);
+        assertEq(merkleRoot, accessCriteria.merkleRoot);
+        assertEq(allowedAddresses, accessCriteria.allowedAddresses);
+    }
 
     // -------------------------------------------------------------------------
     // Test: Internal Functions
+
+    // -------------------------------------------------------------------------
+    // Helper Functions
```
Comment (fabianschu) [inline test/modules/logicModule/LM_PC_FundingPot_v1.t.sol]: While our code of conduct isn't mentioning this specifically, I think the default values (`roundStart` ...) should be placed in the top of the file.

Also I think we shouldn't use wrappers just to call the SuT.

So I think we could remove:
- _helper_createDefaultFundingRound
- _helper_callCreateRound
- _helper_callEditRound
- _helper_createEditedRoundParams

_helper_createAccessCriteria could reuse the default values defined in the top then

File: test/modules/logicModule/LM_PC_FundingPot_v1.t.sol
Comment (fabianschu) [reply test/modules/logicModule/LM_PC_FundingPot_v1.t.sol:1]: There is a compilation error (stack too deep) for this file.

File: test/modules/logicModule/LM_PC_FundingPot_v1.t.sol [hunk 19]
```diff
@@ -95,16 +92,1004 @@ contract LM_PC_FundingPot_v1_Test is ModuleTest {
...
+    │       └── Then it should revert
+    ├── Given hook contract is set
+    │   ├── And hook function is empty
+    │   └── When user attempts to edit the round
+    │       └── Then it should revert
+    └── Given hook function is set
+        ├── And hook contract is empty
+            └── When user attempts to edit the round
+                └── Then it should revert  
+    */
```
Comment (fabianschu) [inline test/modules/logicModule/LM_PC_FundingPot_v1.t.sol]:
```suggestion
    ├── Given user does not have FUNDING_POT_ADMIN_ROLE
    │   └── When user attempts to edit a round
    │       └── Then it should revert
    │
    ├── Given round does not exist
    │   └── When user attempts to edit the round
    │       └── Then it should revert
    │
    ├── Given round is active
    │   └── When user attempts to edit the round
    │       └── Then it should revert
    │
    ├── Given round start time is in the past
    │   └── When user attempts to edit a round with the above parameter
    │       └── Then it should revert
    │
    ├── Given round end time == 0
    ├── And round cap == 0
    │   └── When user attempts to edit a round with the above parameters
    │       └── Then it should revert
    │
    ├── Given round end time is set
    ├── And round end is before round start
    │   └── When user attempts to edit the round
    │       └── Then it should revert
    │
    ├── Given hook contract is set
    ├── And hook function is empty
    │   └── When user attempts to edit the round
    │       └── Then it should revert
    │
    ├── Given hook function is set
    └── And hook contract is empty
        └── When user attempts to edit the round
            └── Then it should revert
```

**Review by fabianschu — CHANGES_REQUESTED** [review]
Great job :) I left a bunch of comments, nothing structural.

File: test/modules/logicModule/LM_PC_FundingPot_v1.t.sol (hunk 18 above)
Comment (zzzuhaibmohd) [reply test/modules/logicModule/LM_PC_FundingPot_v1.t.sol]: @fabianschu we are planning to fuzz these values in the next PR review!

File: src/modules/logicModule/interfaces/ILM_PC_FundingPot_v1.sol (hunk 7 above)
Comment (zzzuhaibmohd) [reply src/modules/logicModule/interfaces/ILM_PC_FundingPot_v1.sol:36]: @fabianschu can you explain the above in detail?

File: src/modules/logicModule/interfaces/ILM_PC_FundingPot_v1.sol (hunk 3 above)
Comment (leeftk) [reply src/modules/logicModule/interfaces/ILM_PC_FundingPot_v1.sol]: We should def change this nice callout.

**Comment by leeftk** [comment]
Realizing I never commented on this last night I just pinged Zuhaib internally. Anyway, fixes LGTM let us know about the extension y'all use for the gherkin so we can start formatting that properly as well! Thanks :)

File: src/modules/logicModule/LM_PC_FundingPot_v1.sol [hunk 20]
```diff
@@ -97,9 +127,249 @@ contract LM_PC_FundingPot_v1 is
...
+            bytes32 merkleRoot,
+            address[] memory allowedAddresses
+        )
+    {
+        Round storage round = rounds[roundId_];
+        AccessCriteria storage accessCriteria = round.accessCriterias[id_];
+
+        isOpen = (
+            accessCriteria.accessCriteriaType == AccessCriteriaType.UNSET
+                || accessCriteria.accessCriteriaType == AccessCriteriaType.OPEN
```
Comment (fabianschu) [inline src/modules/logicModule/LM_PC_FundingPot_v1.sol]: This seems like unintended behaviour. Why is the access criteria considered "open" if no access criteria type is set?

File: src/modules/logicModule/LM_PC_FundingPot_v1.sol [hunk 21]
```diff
@@ -97,9 +127,249 @@ contract LM_PC_FundingPot_v1 is
...
+        }
+
+        if (round_.hookContract == address(0) && round_.hookFunction.length > 0)
+        {
+            revert
+                Module__LM_PC_FundingPot__HookContractRequiredWithHookFunction();
+        }
+    }
+
+    /// @notice Validates the round parameters.
```
Comment (fabianschu) [inline src/modules/logicModule/LM_PC_FundingPot_v1.sol]: The NatSpec is the same as in l. 328. I think it should tell the dev what the difference to the other validation function is.

File: src/modules/logicModule/interfaces/ILM_PC_FundingPot_v1.sol (hunk 7 above)
Comment (fabianschu) [reply src/modules/logicModule/interfaces/ILM_PC_FundingPot_v1.sol:36]: Sorry, maybe I'm thinking to far ahead, but I would assume if the allowlist is stored as an array, then later on when someone that is on the allowlist wants to contribute, within the contribution function it will be necessary to iterate over the array. This means for each contributor this iteration will need to take place, and contributing is probably by far the most common operation that will take place on this contract.

If on the other hand the allowlist is stored as a mapping during configuration, then it is not necessary to iterate over an array for each contribution and since this is supposed to happen on ETH mainnet, it is maybe more efficient gas-wise. Wdyt?

File: src/modules/logicModule/interfaces/ILM_PC_FundingPot_v1.sol (hunk 4 above)
Comment (fabianschu) [reply src/modules/logicModule/interfaces/ILM_PC_FundingPot_v1.sol]: I think this one was overseen :)

File: src/modules/logicModule/interfaces/ILM_PC_FundingPot_v1.sol (hunk 10 above)
Comment (fabianschu) [reply src/modules/logicModule/interfaces/ILM_PC_FundingPot_v1.sol]: I think this might have been overlooked at this place :)

File: test/modules/logicModule/LM_PC_FundingPot_v1.t.sol
Comment (fabianschu) [reply test/modules/logicModule/LM_PC_FundingPot_v1.t.sol:1]: Noticing that in testSupportsInterface there is a duplicate assertion

File: test/modules/logicModule/LM_PC_FundingPot_v1.t.sol [hunk 22]
```diff
@@ -95,16 +90,1197 @@ contract LM_PC_FundingPot_v1_Test is ModuleTest {
...
+        └── Given all the valid parameters are provided
+            └── When user attempts to set access criteria
+                └── Then it should not revert
+    */
+
+    function testFuzzSetAccessCriteria_revertsGivenUserDoesNotHaveFundingPotAdminRole(
+        uint8 accessCriteriaEnum_,
+        address user_
+    ) public {
+        vm.assume(accessCriteriaEnum_ >= 0 && accessCriteriaEnum_ <= 3);
```
Comment (fabianschu) [inline test/modules/logicModule/LM_PC_FundingPot_v1.t.sol]: I don't understand why this uses a different vm.assume than the other fuzz tests. Could you explain your intentions?

File: test/modules/logicModule/LM_PC_FundingPot_v1.t.sol [hunk 23]
```diff
@@ -95,16 +90,1197 @@ contract LM_PC_FundingPot_v1_Test is ModuleTest {
...
+    }
 
     // -------------------------------------------------------------------------
     // Test: Internal Functions
+
+    // -------------------------------------------------------------------------
+    // Helper Functions
+
+    // @notice Creates a default funding round
+    function _helper_createDefaultFundingRound()
```
Comment (fabianschu) [inline test/modules/logicModule/LM_PC_FundingPot_v1.t.sol:1106]: Now that you have added fuzz tests, I still think this test would be easiert to read if the default values were stored on top of the file. Could we change this?

File: test/modules/logicModule/LM_PC_FundingPot_v1.t.sol [hunk 24]
```diff
@@ -95,16 +90,1197 @@ contract LM_PC_FundingPot_v1_Test is ModuleTest {
...
+            roundCap,
+            hookContract,
+            hookFunction,
+            autoClosure,
+            globalAccumulativeCaps
+        );
+    }
+
+    // @notice calls the create round function
+    function _helper_callCreateRound(
```
Comment (fabianschu) [inline test/modules/logicModule/LM_PC_FundingPot_v1.t.sol]: I still don't understand the case for wrappers like this. If it is just calling the function on the funding pot, why not call it directly and remove that abstraction layer?

File: test/modules/logicModule/LM_PC_FundingPot_v1.t.sol [hunk 25]
```diff
@@ -95,16 +90,1197 @@ contract LM_PC_FundingPot_v1_Test is ModuleTest {
...
+            roundCap,
+            hookContract,
+            hookFunction,
+            autoClosure,
+            globalAccumulativeCaps
+        );
+    }
+
+    // @notice Creates a predefined funding round with edited parameters for testing
+    function _helper_createEditedRoundParams()
```
Comment (fabianschu) [inline test/modules/logicModule/LM_PC_FundingPot_v1.t.sol]: I think this helper obscures what is happening in the tests. Instead of using it, could you store the default values on top of the file and when you actually want to edit round parameters in the test, make it clear in each test which parameter is being edited? 

Example:
In `testEditRound_revertsGivenUserIsNotFundingPotAdmin` you are calling this helper to test that only admin can call it. If you had the default values on top of the file you could call the editRound function directly with the default values and test the same thing. 

If you would like to test that editing reverts if you are trying to set the new start date after the end date, then you could just exchange the startDate with an invalid one etc.

Does that make sense?

File: test/modules/logicModule/LM_PC_FundingPot_v1.t.sol [hunk 26]
```diff
@@ -95,16 +90,1197 @@ contract LM_PC_FundingPot_v1_Test is ModuleTest {
...
+            roundCap_,
+            hookContract_,
+            hookFunction_,
+            autoClosure_,
+            globalAccumulativeCaps_
+        );
+    }
+
+    // @notice calls the create round function
+    function _helper_callEditRound(
```
Comment (fabianschu) [inline test/modules/logicModule/LM_PC_FundingPot_v1.t.sol]: Let's remove this wrapper also?

File: test/modules/logicModule/LM_PC_FundingPot_v1.t.sol [hunk 27]
```diff
@@ -95,16 +90,1197 @@ contract LM_PC_FundingPot_v1_Test is ModuleTest {
...
+            roundEnd,
+            roundCap,
+            hookContract,
+            hookFunction,
+            autoClosure,
+            globalAccumulativeCaps
+        );
+    }
+
+    function _helper_createAccessCriteria(uint8 accessCriteriaEnum)
```
Comment (fabianschu) [inline test/modules/logicModule/LM_PC_FundingPot_v1.t.sol:1134]: This helper makes sense to me :)

File: test/modules/logicModule/LM_PC_FundingPot_v1.t.sol [hunk 28]
```diff
@@ -95,16 +90,1197 @@ contract LM_PC_FundingPot_v1_Test is ModuleTest {
...
+                    address(0x0),
+                    bytes32(uint(0x0)),
+                    allowedAddresses
+                );
+            }
+        }
+    }
+
+    // Helper function to set up a round with access criteria
+    function _helper_setupRoundWithAccessCriteria(uint8 accessCriteriaEnum)
```
Comment (fabianschu) [inline test/modules/logicModule/LM_PC_FundingPot_v1.t.sol:1193]: Also makes sense to me

**Review by fabianschu — CHANGES_REQUESTED** [review]
Found some more things mostly related to the clarity of the tests.

File: test/modules/logicModule/LM_PC_FundingPot_v1.t.sol (hunk 18 above)
Comment (fabianschu) [reply test/modules/logicModule/LM_PC_FundingPot_v1.t.sol]: I added more comments on this. I still think these helpers are redundant and removing them would improve the clarity of the tests.

File: test/modules/logicModule/LM_PC_FundingPot_v1.t.sol (hunk 22 above)
Comment (leeftk) [reply test/modules/logicModule/LM_PC_FundingPot_v1.t.sol:804]: You're right these are unecessary. Removing these, nice catch.

File: src/modules/logicModule/LM_PC_FundingPot_v1.sol (hunk 20 above)
Comment (leeftk) [reply src/modules/logicModule/LM_PC_FundingPot_v1.sol]: yes this is right I removed this.

File: test/modules/logicModule/LM_PC_FundingPot_v1.t.sol (hunk 23 above)
Comment (leeftk) [reply test/modules/logicModule/LM_PC_FundingPot_v1.t.sol:1106]: Yup just added this to the top, I basically created a set of default values and then a set of edited values. Though I do think for some tests like creating a round these values should be fuzzed as we should be testing a bunch of different combinations of possible rounds to make sure there aren't an edge cases that break the creation of a round, however I agree that having these in the set up makes the most sense so Im changing that for now.

File: test/modules/logicModule/LM_PC_FundingPot_v1.t.sol (hunk 22 above)
Comment (leeftk) [reply test/modules/logicModule/LM_PC_FundingPot_v1.t.sol]: Don't think this made sense, I changed all of them to be the same and the tests are still passing so I think this was just a mismatch and not intentional.

**Review by fabianschu — APPROVED** [review]
Looks good to me
